### PR TITLE
Add torch array support for the variable explorer

### LINKED ISSUE #7042: Support Torch.Tensor in the variable explorer
## Problem Description
I'm a pytorch user, and I'm always thinking that type torch.Tensor might be represented in a proper way. What torch.Tensor type data need is just '.numpy()' to be converted to numpy array objects.

So, is it possible to apply simple scripts for the unsupported data types inside of Variable explorer?(i.e. If Variable explorer saw Tensor type objects, then a corresponding function/method for the type is applied and show the result of it on Variable explorer with distinguishable color.)
```python
def type_conversion(data: torch.Tensor):
  if isinstance(data, torch.Tensor):
    return data.numpy()
  return None
```

## Package Versions

* Spyder: 3.2.8
* Python: 3.6.4
* Qt:
* PyQt:
* Operating System: macOS 10.13.4

### Dependencies

```
IPython >=4.0    :  6.3.1 (OK)
cython >=0.21    :  0.28.2 (OK)
jedi >=0.9.0     :  0.12.0 (OK)
nbconvert >=4.0  :  5.3.1 (OK)
numpy >=1.7      :  1.14.1 (OK)
pandas >=0.13.1  :  0.22.0 (OK)
psutil >=0.3     :  5.4.5 (OK)
pycodestyle >=2.3:  2.4.0 (OK)
pyflakes >=0.6.0 :  1.6.0 (OK)
pygments >=2.0   :  2.2.0 (OK)
pylint >=0.25    :  1.8.4 (OK)
qtconsole >=4.2.0:  4.3.1 (OK)
rope >=0.9.4     :  0.10.7 (OK)
sphinx >=0.6.6   :  1.7.4 (OK)
sympy >=0.7.3    :  1.1.1 (OK)
```

### PR DESCRIPTION
**Pull request by dfardila**
## Introduction
Spyder is one of the favorite data science tools for academics and professionals around the world, thanks for its intuitive workflow, integrated debugging capabilities with IPython and its simple but yet powerful variable explorer.

In the recent years, the day-to-day libraries that a data scientist uses has expanded significantly, making the support of more data libraries an important feature for the spyder variable explorer.

With this in mind, I propose a simple solution to add torch arrays support for the variable explorer of spyder.

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
* [x] Included a screenshot or animation.
spyder-variableexplorer-torch

The main strategy to add support for torch arrays consisted in converting the data to a numpy array early in the data verification process. This conversion is implemented in [arrayeditor.py - line 941](https://github.com/spyder-ide/spyder/compare/master...dfardila:spyder:feature-torch-variable-explorer?expand=1#diff-caadc07a1a174b1deabb8b0ba33df93a1e4c930f1418dba5966ccce0d0959808R941).

To verify if the array being explored is a torch array, the torch library needs to be imported in a safe way. This is done in [arrayeditor.py - line 49](https://github.com/spyder-ide/spyder/compare/master...dfardila:spyder:feature-torch-variable-explorer?expand=1#diff-caadc07a1a174b1deabb8b0ba33df93a1e4c930f1418dba5966ccce0d0959808R49-R53) and [collectionsdelegate.py - line 60](https://github.com/spyder-ide/spyder/compare/master...dfardila:spyder:feature-torch-variable-explorer?expand=1#diff-b07ce1a81591b4285161bec22f1644b8d55c636a82bdcf62e7fe029dfb95acd4R60-R64). Not sure if the imports should go to the spyder kernels, if so please let me know.

The remaining changes manage the torch array before the conversion. 

I changed the title of the variable explorer to from "Numpy Object Array" to "Object Array", so that it can be used for numpy or torch arrays.

### Issue(s) Resolved

Fixes #7042 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
dfardila

### CODE REVIEW

File: spyder/plugins/variableexplorer/widgets/arrayeditor.py [hunk 1]
```diff
@@ -922,13 +927,23 @@ def set_data_and_check(self, data, readonly=False):
...
+            if torch is not None and isinstance(data, torch.Tensor):
+                pass
+            else:
+                return False
         is_masked_array = isinstance(data, np.ma.MaskedArray)
+        is_torch_tensor = isinstance(data, torch.Tensor)
 
+        if is_torch_tensor:
+            readonly = True
+            data = data.cpu().numpy() #data is converted to numpy array for display and edition, but the original tensor is not modified when accepting changes, so we can keep it read-only
```
Comment (dfardila) [inline spyder/plugins/variableexplorer/widgets/arrayeditor.py]: Data is converted from torch.tensor to np.array. This is the main strategy to add support and reuse the already implemented features.

File: spyder/plugins/variableexplorer/widgets/arrayeditor.py [hunk 2]
```diff
@@ -938,7 +938,7 @@ def set_data_and_check(self, data, readonly=False):
 
         if is_torch_tensor:
             readonly = True
-            data = data.cpu().numpy() #data is converted to numpy array for display and edition, but the original tensor is not modified when accepting changes, so we can keep it read-only
+            data = data.detach().cpu().numpy() #data is converted to numpy array for display, but the original tensor is not modified when accepting changes, so we can keep it read-only
```
Comment (dfardila) [inline spyder/plugins/variableexplorer/widgets/arrayeditor.py:966]: Added detach() to make sure that the torch tensor is not linked to the gradient calculation graph.

**Comment by ccordoba12** [comment]
Thanks for your contribution @dfardila! You said:

> The main strategy to add support for torch arrays consisted in converting the data to a numpy array early in the data verification process. This conversion is implemented in [arrayeditor.py - line 941](https://github.com/spyder-ide/spyder/compare/master...dfardila:spyder:feature-torch-variable-explorer?expand=1#diff-caadc07a1a174b1deabb8b0ba33df93a1e4c930f1418dba5966ccce0d0959808R941).

Unfortunately, this is not a good strategy because it'll freeze Spyder when trying to explore large torch arrays. In other words, Spyder will stop working completely until the conversion is fully done, which wouldn't be a nice experience for our users.

Instead, the conversion needs to be done on the kernel side, i.e. before sending a serialized copy of the object from spyder-kernels to Spyder. Since a kernel is a Python interpreter started in a separate process than Spyder, any computation taken place on it doesn't freeze Spyder itself.

To understand how to implement that, please take a look at the way support for Polars dataframes was added in pull request #24558. And since you'll need to make a joint pull request in spyder-kernels to support torch arrays, please follow the instructions in our [Contributing guide](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md#making-contributions-that-depend-on-pull-requests-in-spyder-kernels) on how to do include your pull request there in a Spyder one.

Let us know if you additional questions about that process or the implementation itself.
